File: service/src/main/java/com/service/service/KrBankApiService.java
package com.service.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.service.config.KrBankProperties;
import com.service.domain.KrBankData;
import com.service.domain.KrBankSchema;
import com.service.dto.KrBankDataResponse;
import com.service.dto.KrBankRequest;
import com.service.dto.KrBankSchemaResponse;
import com.service.repository.KrBankDataRepository;
import com.service.repository.KrBankSchemaRepository;
import com.google.gson.Gson;

import java.net.URI;
import java.util.List;
import java.util.Map;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;


@Slf4j
@Service
public class KrBankApiService {
    @Autowired
    Gson gson;

    @Autowired
    RestTemplate restTemplate;

    @Autowired
    KrBankProperties krBankProperties;

    @Autowired
    KrBankDataRepository krBankDataRepository;

    @Autowired
    KrBankSchemaRepository krBankSchemaRepository;

    List<KrBankData> saveData(List<KrBankData> krBankDatas) {
        krBankDatas.forEach(i -> krBankDataRepository.save(i));
        return krBankDatas;
    }

    List<KrBankSchema> saveSchema(List<KrBankSchema> krBankSchemas) {
        krBankSchemas.forEach(i -> krBankSchemaRepository.save(i));
        return krBankSchemas;
    }

    public List<KrBankData> getDataFromAPI(KrBankRequest krBankRequest) {
        krBankRequest.setServiceName("StatisticSearch");
        krBankRequest.setAuthKey(krBankProperties.getApikey());
        krBankRequest.setPeriod("DD");
//        ResponseEntity<String> response = restTemplate.getForEntity(krBankRequest.getUrlString(), String.class);
//        KrBankDataResponse krBankDataResponse = gson.fromJson(response.getBody(), KrBankDataResponse.class);

        List<KrBankData> krBankDatas = gson.fromJson(
                restTemplate.getForEntity(getUrlString(krBankRequest), String.class).getBody(),
                KrBankDataResponse.class
        ).getStatisticSearch().getRow();
        return krBankDatas;
    }

    public List<KrBankSchema> getSchemaFromAPI() {
        KrBankRequest krBankRequest = new KrBankRequest();
        krBankRequest.setServiceName("StatisticTableList");
        krBankRequest.setAuthKey(krBankProperties.getApikey());

        ResponseEntity<String> response = restTemplate.getForEntity(getUrlString(krBankRequest), String.class);
        KrBankSchemaResponse krBankSchemaResponse = gson.fromJson(response.getBody(), KrBankSchemaResponse.class);
        return krBankSchemaResponse.getStatisticTableList().getKrBankSchema();
    }

    public List<KrBankData> batchData(KrBankRequest krBankRequest) {
        return saveData(getDataFromAPI(krBankRequest));
    }

    public List<KrBankSchema> batchSchema() {
        return saveSchema(getSchemaFromAPI());
    }


    //    http://ecos.bok.or.kr/api/StatisticTableList/sample/xml/kr/1/10/
    //    http://ecos.bok.or.kr/api/StatisticSearch/sample/xml/kr/1/10/010Y002/MM/201101/201101/AAAA11/
    public URI getUrlString(KrBankRequest krBankRequest) {
        String retString = krBankRequest.getUrl() + "/{serviceName}/{authKey}/{requestType}/{language}/{reqStartCount}/{reqEndCount}" +
                "/{statisticCode}/{period}/{queryStartDate}/{queryEndDate}";

//        Map<String, Object> params = new HashMap<String, Object>();

        ObjectMapper objectMapper = new ObjectMapper();
        Map result = objectMapper.convertValue(krBankRequest, Map.class);


        return UriComponentsBuilder.fromUriString(retString)
                .buildAndExpand(result)
                .toUri();

//        params.put("authKey", this.authKey);
//        params.put("requestType", this.requestType);
//        params.put("language", this.language);
//        params.put("statisticCode", this.statisticCode);
//        params.put("option1", this.option1);
//        params.put("option2", this.option2);
//        params.put("option3", this.option3);
//        params.put("queryStartDate", this.queryStartDate);
//        params.put("queryEndDate", this.queryEndDate);
//        params.put("reqStartCount", this.reqStartCount);
//        params.put("reqEndCount", this.reqEndCount);
//        params.put("period", this.period);

//        return retString;
    }

}

File: src/main/kotlin/com/minssan9/financial/krbank/service/KrBankApiService.kt
package com.minssan9.financial.krbank.service

import com.google.gson.Gson
import com.minssan9.financial.config.AppProperties
import com.minssan9.financial.krbank.domain.KrBankData
import com.minssan9.financial.krbank.domain.KrBankSchema
import com.minssan9.financial.krbank.dto.KrBankDto
import com.minssan9.financial.krbank.repository.KrBankDataRepository
import com.minssan9.financial.krbank.repository.krBankSchemaRepository
import lombok.extern.slf4j.Slf4j
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import org.springframework.web.client.getForEntity


@Slf4j
@Service
@ConfigurationProperties("account")
class KrBankApiService {
    @Autowired
    lateinit var gson : Gson


    @Autowired
    lateinit var restTemplate : RestTemplate

    @Autowired
    lateinit var appProperties: AppProperties

    @Autowired
    lateinit var krBankDataRepository: KrBankDataRepository

    @Autowired
    lateinit var krBankSchemaRepository: krBankSchemaRepository


    fun saveData(krBankDatas: List<KrBankData>) {
        krBankDatas.forEach { i -> krBankDataRepository.save(i) }
    }

    fun saveSchema(krBankSchemas: List<KrBankSchema>) {
        krBankSchemas.forEach { i -> krBankSchemaRepository.save(i) }
    }

    fun getDataFromAPI(krBankRequest: KrBankDto.KrBankRequest): List<KrBankData>? {
        val apiDto =KrBankDto()

        krBankRequest.serviceName = "StatisticSearch"

        var response: ResponseEntity<String> = restTemplate.getForEntity<String>(apiDto.getUrlString(krBankRequest))!!
        var krBankDataResponse: KrBankDto.KrBankDataResponse =
            gson?.fromJson(response.body, KrBankDto.KrBankDataResponse::class.java)

        return krBankDataResponse.krBankResult.row
    }

    fun getSchemaFromAPI(krBankRequest: KrBankDto.KrBankRequest ): List<KrBankSchema>? {
        val apiDto =KrBankDto()


        var krBankRequest = KrBankDto.KrBankRequest("","","","","","",krBankRequest.reqStartCount,krBankRequest.reqEndCount)
        krBankRequest.serviceName = "StatisticTableList"

        var response: ResponseEntity<String> = restTemplate.getForEntity<String>(apiDto.getUrlString(krBankRequest))!!
        var krBankSchemaResponse: KrBankDto.KrBankSchemaResponse =
            gson?.fromJson(response.body, KrBankDto.KrBankSchemaResponse::class.java)

        return krBankSchemaResponse.krBankResult.row
    }

    fun getKOSPI(queryStartDate: String, queryEndDate: String, startPos: Long, endPos: Long): List<KrBankData>? {
        return getDataFromAPI(
            krBankRequest =
            KrBankDto.KrBankRequest(
                "064Y001",
                "0001000",
                "",
                "",
                queryStartDate,
                queryEndDate,
                startPos, endPos
            )
        )
    }

}
